promql/functions.go

```diff
@@ -184,6 +184,107 @@ func funcIrate(ev *evaluator, args Expressions) model.Value {
 	return resultVector
 }
 
+// doubleSVal calculates the smoothed value at index i of raw data d
+// s = computed smoothed values, b = computed trend factors, d = raw
+func doubleSVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
+	// check for the cached value
+	if f := s[i]; !math.IsNaN(f) {
+		return f
+	}
+
+	x := sf * d[i]
+	y := (1 - sf) * (doubleSVal(i-1, sf, tf, s, b, d) + doubleBVal(i-1, sf, tf, s, b, d))
+
+	// cache value
+	s[i] = x + y
+
+	return s[i]
+}
+
+// doubleBVal calculates the trend value at index i of raw data d
+// s = computed smoothed values, b = coumputed trend factors, d = raw
+func doubleBVal(i int, sf, tf float64, s, b, d []float64) float64 {
+
+	// check for the cached value
+	if f := b[i]; !math.IsNaN(f) {
+		return f
+	}
+
+	x := tf * (doubleSVal(i, sf, tf, s, b, d) - doubleSVal(i-1, sf, tf, s, b, d))
+	y := (1 - tf) * doubleBVal(i-1, sf, tf, s, b, d)
+
+	// cache value
+	b[i] = x + y
+
+	return b[i]
+}
+
+// === holt_winters(node model.ValVector, smoothingFactor model.ValScalar, trendFactor model.ValScalar) Vector ===
+// algorhythm taken from https://en.wikipedia.org/wiki/Exponential_smoothing
+func funcHoltWinters(ev *evaluator, args Expressions) model.Value {
+	mat := ev.evalMatrix(args[0])
+
+	// smoothing factor
+	sf := ev.evalFloat(args[1])
+
+	// trend factor
+	tf := ev.evalFloat(args[2])
+
+	// make an ouput vector large enough to hold the entire result
+	resultVector := make(vector, len(mat.value()))
```

Comment:
I think you want `make(vector, 0, len(mat.value()))` and then you can remove the `resultCount` logic.